frontend/packages/aquascope-editor/src/editor-utils/stepper.tsx

```diff
@@ -0,0 +1,383 @@
+        ".step-widget-contents"
+      );
+      let btn = container.querySelector<HTMLSpanElement>(
+        ".step-table-collapse"
+      );
+
+      if (contents) {
+        contents.style.width = this.tableHidden ? "0px" : "auto";
+      }
+      if (btn) {
+        btn.innerHTML = this.tableHidden ? arrowOut : arrowIn;
+      }
+    };
+    let matchAll = new RegExp("(.*)?");
+    let r = this.focusedPaths.get(currLine.number) ?? matchAll;
+    let [focusedDiffs, hiddenDiffs] = partition(
+      ([path, _]) => !!path.match(r),
+      this.step.state
+    );
+
+    let hiddenDropdown =
+      hiddenDiffs.length > 0 ? (
+        <>
+          <div className="step-table-dropdown">...</div>
+          <div className="step-table-content">
+            <table className="perm-step-table">
+              {hiddenDiffs.map(([path, diffs]) => (
+                <PermDiffRow path={path} diffs={diffs} />
+              ))}
+            </table>
+          </div>
+        </>
+      ) : null;
+
+    let rowStyle = {
+      width: hideTable ? "0px" : "auto",
+    };
+
+    ReactDOM.createRoot(container).render(
+      <div className="perm-step-widget">
+        <span className="step-table-collapse" onClick={toggleWidget}>
+          {hideTable ? arrowOut : arrowIn}
+        </span>
+        <div className="step-widget-contents" style={rowStyle}>
+          {spaces}
+          <div className="step-table-container" onClick={toggleHiddenPaths}>
+            <div className="step-table-parent">
+              <table className="perm-step-table">
+                {focusedDiffs.map(([path, diffs]) => (
+                  <PermDiffRow path={path} diffs={diffs} />
+                ))}
+              </table>
+              {hiddenDropdown}
+            </div>
+          </div>
+        </div>
+      </div>
+    );
+
+    return container;
+  }
+
+  ignoreEvent(): boolean {
+    return true;
+  }
+}
+
+export function renderSteps(
+  view: EditorView,
+  container: HTMLDivElement,
+  stateSteps: PermissionsStateStep[],
+  config: StepperConfig
+) {
+  // Normalize the config
+  let doc = view.state.doc;
+  let focusedPaths = new Map<number, RegExp>();
+  let focusedLines =
+    !config.focusedCharPos || config.focusedCharPos.length == 0
+      ? Array.from({ length: doc.lines }, (value, key) => key + 1)
+      : config.focusedCharPos!.map(pos => doc.lineAt(pos).number);
+
+  // Insert the regex for a specified char pos
+  config.focusedPaths?.forEach((re: string, pos: number) => {
+    let line = doc.lineAt(pos).number;
+    focusedPaths.set(line, new RegExp(re));
+  });
+
+  let decos = _.sortBy(
+    stateSteps.map(step =>
+      Decoration.widget({
+        widget: new PermissionStepTableWidget(
+          view,
+          config,
+          step,
+          focusedLines,
+          focusedPaths
+        ),
+      }).range(stepLocation(step))
+    ),
+    deco => deco.from
+  );
+
+  let display = true;
+  let plugin = ViewPlugin.fromClass(
```

Comment:
Could you make `display` a class member instead of a capture from the environment? That would let you define this plugin once rather than every time you render.